api/api.go

```diff
@@ -921,32 +924,20 @@ func (api *Server) getBlockMetas(start uint64, count uint64) (*iotexapi.GetBlock
 	}
 	var res []*iotextypes.BlockMeta
 	for height := start; height <= tipHeight && count > 0; height++ {
-		blk, err := api.bc.GetBlockByHeight(height)
-		if err != nil {
-			return nil, status.Error(codes.NotFound, err.Error())
-		}
-
-		hash := blk.HashBlock()
-		ts, _ := ptypes.TimestampProto(blk.Header.Timestamp())
-		txRoot := blk.TxRoot()
-		receiptRoot := blk.ReceiptRoot()
-		deltaStateDigest := blk.DeltaStateDigest()
-		transferAmount := getTranferAmountInBlock(blk)
-		logsBloom := blk.Header.LogsBloomfilter()
-
-		blockMeta := &iotextypes.BlockMeta{
-			Hash:             hex.EncodeToString(hash[:]),
-			Height:           blk.Height(),
-			Timestamp:        ts,
-			NumActions:       int64(len(blk.Actions)),
-			ProducerAddress:  blk.ProducerAddress(),
-			TransferAmount:   transferAmount.String(),
-			TxRoot:           hex.EncodeToString(txRoot[:]),
-			ReceiptRoot:      hex.EncodeToString(receiptRoot[:]),
-			DeltaStateDigest: hex.EncodeToString(deltaStateDigest[:]),
-		}
-		if logsBloom != nil {
-			blockMeta.LogsBloom = hex.EncodeToString(logsBloom.Bytes())
+		var blockMeta *iotextypes.BlockMeta
```

Comment:
Almost good. Please see if it's better to refactor by following this pseudo code:

```golang
blkMeta, err := getBlockMetaByHeader(height)
if errors.Cause(err) == db.NotExist && (db.Error() == "missing numActions" || db.Error == ""missing transferAmount") {
    blkMeta, err := getBlockMetaByBlock(height)
}

...


func getBlockMetaByBlock(height) (BlockMeta, error) {
    // copy of existing code
    ...
}

func getBlockMetaByHeader(height) (BlockMeta, error) {
    // new logic of getting block meta by header
    ...
    numActions, err := ...
    if err != null {
        return nil, errors.Wrap(err, "missing numActions")
    }
    ...
    transferAmount, err := ...
    if err != null {
        return nil, errors.Wrap(err, "missing transferAmount")
    }
    ...
}
```

Reply:
Could you specify the meaning of db.Error()? Is err.Error()? The error I can't find in db package.

Reply:
Sorry my bad.

db.Error() -> err.Error(). It's the interface of error

And check https://godoc.org/github.com/pkg/errors#Cause for how to use errors.Cause. Also, search for the examples in our codebase.

Reply:
Should we only checks `errors.Cause(err) == db.NotExist`.
I tried to implement above code and found that both err.Error() and errors.Cause(err).Error() aren't worked as well for if.
Our codebase example also only has the `errors.Cause(err) == db.NotExist` check.